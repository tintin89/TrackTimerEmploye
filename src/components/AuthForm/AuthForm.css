.formAuth{
    border-radius: 3px;
    box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.5);
    max-width:500px;
    width: 90%;
    background-color: #1c1f24;
    display: flex;
    flex-flow: column;
    align-items: center;
    padding: 0.5rem;   
    font-family: 'Poppins', sans-serif;
}

.formAuth__field{
    margin-top: 1rem;
    width: 100%;    
    display: flex;
    flex-flow: column;
    align-items: center;
}

.formAuth__field input{
    width: 90%;
    padding: 0.5rem;
    outline: none;
    border: none;
    color: #3d3d3d;
    border-radius: 3px;
}

.formAuth__field input::placeholder{
  color: #686868;
}

.formAuth__field label{
    width: 95%;
    padding:0.5rem 0;
    color: #ffff;
}

.formAuth button{
    cursor: pointer;
    background-color: #2e6474;
    color: #ffff;
    border: none;
    outline: none;
    border-radius: 3px;
    margin-top: 2rem;
    width: 70%;
    padding:0.3rem;
}

.formAuth__register{
    cursor: pointer;
    width: 90%;
    margin-top: 2rem;
    color: #ffff;
    text-align: end;
    
}