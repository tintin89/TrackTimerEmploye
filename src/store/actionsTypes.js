export const UPDATE_USER_INFO = 'UPDATE_USER_INFO';
export const UPDATE_USER_DATA = 'UPDATE_USER_DATA';
export const UPDATE_ARRIVING_HOUR = "UPDATE_ARRIVING_HOUR";
export const UPDATE_ERROR_M = 'UPDATE_ERROR_M';
export const UPDATE_LOADING = 'UPDATE_LOADING';
export const UPDATE_LUNCHING = 'UPDATE_LUNCHING';
export const UPDATE_LUNCH_BREAK = 'UPDATE_LUNCH_BREAK';


