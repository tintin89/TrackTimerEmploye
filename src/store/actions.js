import * as actionTypes from './actionsTypes';
import axios from 'axios';
import { apiKey } from '../firebase';
import moment from 'moment';




export const updateArrivingHour = (hour)=>{
    return {
        type:actionTypes.UPDATE_ARRIVING_HOUR,
        payload:hour
    }
}

export const updateLunchBreak = (dataLB)=>{
    return {
        type:actionTypes.UPDATE_LUNCH_BREAK,
        payload:dataLB
    }
}

export const updateIsLunching = (lunching) =>{
    return {
        type:actionTypes.UPDATE_LUNCHING,
        payload:lunching
    }
}

export const updateUserData = (data) =>{
    return {
        type:actionTypes.UPDATE_USER_DATA,
        payload:data
    }
}
export const updateUserInfo=(info)=>{
    return {
        type:actionTypes.UPDATE_USER_INFO,
        payload:info
    }
}

export const updateErrorM = (m) =>{
    return {
        type:actionTypes.UPDATE_ERROR_M,
        payload:m
    }
}


export const updateLoading = (loading) =>{
    return {
        type:actionTypes.UPDATE_LOADING,
        payload:loading
    }
}

export const sign = (email,password,register)=>{
    return dispatch => {
        dispatch(updateLoading(true));
        if(register){           
            axios.post(`https://identitytoolkit.googleapis.com/v1/accounts:signUp?key=${apiKey}`,
            {email:email,password:password,returnSecureToken:true})
            .then(async response=>{
               
                dispatch(updateUserInfo(response.data));
                  localStorage.setItem("userInfo",JSON.stringify(response.data)); 
              try{
                  const date = moment();
              await axios.put(`https://timertracker-a17c6-default-rtdb.firebaseio.com/employers/${response.data.localId}.json`,
                {
                    email:response.data.email,
                   [date.format("MMM Do YY")]:{                   
                    arrivingHour:date.format('LT'),
                    exitingHour:"",
                    workedHours:0,
                    lunchBreaks:[]
                   }
                })  
                dispatch(updateLoading(false));     
                dispatch(updateArrivingHour(date.format('LT')));
              }   
              catch(error){
                  console.log("errorDB",error);
              }           
            })
            .catch(error=>{
                dispatch(updateErrorM(error.response.data.error.message));
               dispatch(updateLoading(false));
            })
        }else{            
            axios.post(`https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${apiKey}`,
            {email:email,password:password,returnSecureToken:true})
            .then(async response=>{                
                dispatch(updateUserInfo(response.data));
                localStorage.setItem("userInfo",JSON.stringify(response.data));
                try {
               const data = await axios.get(`https://timertracker-a17c6-default-rtdb.firebaseio.com/employers/${response.data.localId}.json`);
               if(data.data[moment().format("MMM Do YY")]===undefined){               
                localStorage.setItem("arrivingHour",moment().format('LT'));
                dispatch(updateArrivingHour(moment().format('LT')));
               }else{
                dispatch(updateArrivingHour(data.data[moment().format("MMM Do YY")].arrivingHour));
               }
                dispatch(updateUserData(data.data[moment().format("MMM Do YY")].workedHours));
                
                localStorage.setItem("existCounter",true);
               
               

                } catch (error) {
                    
                }
                dispatch(updateLoading(false));
            })
            .catch(error=>{
                dispatch(updateErrorM(error.response.data.error.message));
                dispatch(updateLoading(false));
            })
        }
    }
}

export const logout = ()=>{
    return dispatch=>{
        localStorage.clear();
        dispatch(updateUserInfo(""));
        dispatch(updateUserData(""));
        dispatch(updateArrivingHour(""));
        
        
    }
}


