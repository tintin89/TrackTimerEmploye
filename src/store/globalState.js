import * as actionType from './actionsTypes';


const initialState = {
    userData:localStorage.getItem('userData') ? JSON.parse(localStorage.getItem('userData')) : "",
    arrivingHour:localStorage.getItem('arrivingHour') ? JSON.parse(localStorage.getItem('arrivingHour')) : "",
    userInfo:localStorage.getItem('userInfo') ? JSON.parse(localStorage.getItem('userInfo')) : "",    
    errorM:"",
    isloading:false,
  
    
}

const globalState = (state=initialState,action)=>{
    switch (action.type) {
       case actionType.UPDATE_USER_INFO:
           return {
               ...state,
               userInfo:action.payload
           }

       
       case actionType.UPDATE_ERROR_M:
            return {
                ...state,
                errorM:action.payload
            }   
            
        case actionType.UPDATE_LOADING:
            return {
                ...state,
                isloading:action.payload
            }   
            
        case actionType.UPDATE_LUNCHING:
            return {
                ...state,
                isLunching:action.payload
            }    

        case actionType.UPDATE_LUNCH_BREAK:
            return {
                ...state,
                lunchBreak:action.payload
            }            

        case actionType.UPDATE_USER_DATA:
            return {
                ...state,
                userData:action.payload
            }  
            
        case actionType.UPDATE_ARRIVING_HOUR:
            return {
                ...state,
                arrivingHour:action.payload
            }    

        default:
            return state
    }
}

export default globalState;