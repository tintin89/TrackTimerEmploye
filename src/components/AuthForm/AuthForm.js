import React,{useState} from 'react';
import './AuthForm.css';
import {useSelector,useDispatch} from 'react-redux';
import {Redirect} from 'react-router-dom';
import {sign,updateErrorM} from '../../store/actions';
import LabelMessage from '../LabelMessage/LabelMessage';
import Loader from '../Loader/Loader';


const mapState = (globalState)=>({
    userInfo:globalState.userInfo,
    errorM:globalState.errorM,
    loading:globalState.isloading
})

function AuthForm() {
    const dispatch = useDispatch();
    const {userInfo,errorM,loading} = useSelector(mapState);
    const[form,setForm] = React.useState({user:"",password:""});
    const [passwordConfirm,setPasswordConfirm] = useState("");
    const [register,setRegister] = useState(false);
   
    

    const handleSubmit = e =>{
        e.preventDefault();
        if(form.user===""||form.password===""){
            dispatch(updateErrorM("Empty Fields!"));
            return;
        }
        if(register){
        if(passwordConfirm===form.password){
           dispatch(sign(form.user,form.password,register));
        }else{
            dispatch(updateErrorM("Passwords don't match"));
        }

        }else{
           dispatch(sign(form.user,form.password,register));
        }

    }

    let redireccionar=null;
    if(userInfo!==""){
         redireccionar=<Redirect exact to="/dashboard"/>
        }

    return (
        <>
        {redireccionar}
        {
            loading ? <Loader/> :
            <form className="formAuth" onSubmit={handleSubmit}>
            <div className="formAuth__field">
            <label htmlFor="email">User</label>
            <input
             value={form.user} 
             onChange={e=>{             
                 setForm({...form,user:e.target.value});
                 if(errorM!==""){
                     dispatch(updateErrorM(""));
                 }
             }} 
             type="email" 
             placeholder="user@example.com" 
             id="email"/>                
            </div>
            <div className="formAuth__field">
                  <label htmlFor="password">Password</label>
                  <input
                  value={form.password}
                  onChange={e=>{
                     setForm({...form,password:e.target.value})
                     if(errorM!==""){
                        dispatch(updateErrorM(""));
                    }
                  }}
                   type="password" 
                   placeholder="Enter your password" 
                   id="password"/>
                  </div>
                {
                    register &&
                    <div className="formAuth__field">
               <label htmlFor="password">Confirm Password</label>
                <input
                  value={passwordConfirm}
                  onChange={e=>{
                     setPasswordConfirm(e.target.value);
                     if(errorM!==""){
                        dispatch(updateErrorM(""));
                    }
                  }}
                   type="password" 
                   placeholder="Confirm your password" 
                   id="password"/>
                  </div>  
                } 
    
                {errorM!=="" && <LabelMessage message={errorM}/>}
    
            <button  type="submit">{register ? "SIGN UP" : "SIGN IN"}</button>   
            <div onClick={()=>setRegister(prevState=>!prevState)} className="formAuth__register">{register ? "Back" : "Register Here"}</div>         
           </form>    
        }
       </>     
    )
}

export default AuthForm
