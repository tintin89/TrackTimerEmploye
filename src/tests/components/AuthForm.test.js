import React from 'react';
import AuthForm from '../../components/AuthForm/AuthForm';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router';
import {sign} from '../../store/actions';


jest.mock('../../store/actions',()=>({
    sign:jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);



describe('test in AuthForm',()=>{
    const store = mockStore({
        userinfo:"",
        errorM:"",
        loading:true
    });
    store.dispatch = jest.fn();
    const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
          <AuthForm/>  
        </MemoryRouter>
        
    </Provider> )

  test('must show correctly',()=>{
      expect(wrapper).toMatchSnapshot();
  })

  test('must call handlesubmit',()=>{
    
     wrapper.find('#email').simulate('change',{target:{value:"test@test.com"}});
     wrapper.find('#password').simulate('change',{target:{value:"test"}});

      wrapper.find('form').prop('onSubmit')(
          {preventDefault(){}}
      );
      expect(sign).toHaveBeenCalled();
  })
})