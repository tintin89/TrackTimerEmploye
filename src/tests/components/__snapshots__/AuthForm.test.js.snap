// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in AuthForm must show correctly 1`] = `
Array [
  "",
  <form
    className="formAuth"
    onSubmit={[Function]}
  >
    <div
      className="formAuth__field"
    >
      <label
        htmlFor="email"
      >
        User
      </label>
      <input
        id="email"
        onChange={[Function]}
        placeholder="user@example.com"
        type="email"
        value=""
      />
    </div>
    <div
      className="formAuth__field"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        id="password"
        onChange={[Function]}
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      SIGN IN
    </button>
    <div
      className="formAuth__register"
      onClick={[Function]}
    >
      Register Here
    </div>
  </form>,
]
`;
